File: services/DatabaseService.rb
require "mysql2"
require_relative "../models/Player.rb"
require_relative "../models/PlayerSeasonStats.rb"
require_relative "../models/PlayerSeasonStatsGoaler.rb"
require_relative "../models/PlayerSalarySeason.rb"
require_relative "../models/PoolPlayerValues.rb"
require_relative "../models/Position.rb"
require_relative "../models/Team.rb"

module DatabaseServicePositions
    def insertPositions(positions)
        positions.each do |item|
            position = Position.fromJson(item)

            result = @insertPosition.execute(
                position.code,
                position.abbrev,
                position.fullName,
                position.type
            )
        end
    end
end

module DatabaseServiceTeams
    def insertTeams(teams)
        teams.each do |item|
            team = Team.fromJson(item)

            result = @insertTeam.execute(
                team.id,
                team.name,
                team.venue,
                team.abbreviation,
                team.firstYearOfPlay,
                team.divisionId,
                team.conferenceId,
                team.franchiseId,
                team.active
            )
        end
    end

    def getTeamForId(teamId)
        results = @dbClient.query("SELECT * FROM Teams where id = #{teamId}")
        row = results.each.first
        return Team.fromRow(row)
    end

    def getAllTeams()
        teams = Array.new
        results = @dbClient.query("SELECT * FROM Teams")

        results.each do |row|
            team = Team.fromRow(row)
            teams.push(team)
        end

        return teams
    end
end

module DatabaseServiceRoster
    def clearTeamRosters()
        result = @dbClient.query("TRUNCATE TABLE TeamsPlayers")
    end

    def insertTeamRoster(team, roster)
        roster.each do |player|
            result = @insertTeamPlayers.execute(
                team.id,
                player["id"]
            )
        end
    end

    def getAllRosters()
        roster = Array.new

        results = @dbClient.query("SELECT * FROM TeamsPlayers")

        results.each do |row|
            roster.append(row["playerId"])
        end

        return roster
    end

    def getRosterForTeam(team)
        roster = Array.new
        results = @dbClient.query("SELECT * FROM TeamsPlayers WHERE teamId = #{team.id}")

        results.each do |row|
            roster.append(row["playerId"])
        end

        return roster
    end
end

class DatabaseService
    def initialize
        @dbClient = Mysql2::Client.new(:host => "localhost", :database => "hockeydata", :username => "sleclerc", :password => "sleclerc")
        prepareStatements()
    end

    include DatabaseServicePositions
    include DatabaseServiceTeams
    include DatabaseServiceRoster

    def getAllPlayers()
        players = Array.new

        results = @dbClient.query("SELECT * FROM Players")

        results.each do |row|
            players.append(row["id"])
        end

        return players
    end

    def insertPlayer(item)
        # p item

        birthDate = item["birthDate"]
        birthDateParts = birthDate.split("-")

        birthProvince = item["birthStateProvince"]["default"] unless item["birthStateProvince"].nil?

        result = @insertPlayer.execute(
            item["playerId"],
            item["firstName"]["default"],
            item["lastName"]["default"],
            item["sweaterNumber"],
            birthDateParts[0],
            birthDateParts[1],
            birthDateParts[2],
            item["birthCity"]["default"],
            birthProvince,
            item["birthCountry"],
            item["heightInInches"],
            item["weightInPounds"],
            item["isActive"],
            item["shootsCatches"],
            false, #Field not present anymore item["rookie"],
            item["currentTeamId"],
            item["position"],
            item["headshot"],
            item["heroImage"]
        )
    end

    def getPlayerForId(playerId)
        players = Array.new
        results = @dbClient.query("SELECT * FROM Players WHERE id = #{playerId}")

        results.each do |row|
            player = Player.fromRow(row)
            players.append(player)
        end

        return players.first
    end

    def insertPlayerArchiveStat(playerId, playerStats)
        playerStats.each do |stat|
            @insertPlayerStats.execute(
                playerId,
                stat["season"],
                stat["gamesPlayed"],
                stat["goals"],
                stat["assists"],
                stat["points"],
                stat["shots"],
                nil, #stat["stat"]["hits"], does not exist anymore
                stat["avgToi"].to_f,
                nil, #|||stat["stat"]["shifts"],
                stat["plusMinus"],
                stat["shootingPctg"],
                stat["pim"],
                stat["powerPlayGoals"],
                stat["powerPlayPoints"],
                nil, #!!!convertStringToTime(stat["stat"]["powerPlayTimeOnIce"]),
                stat["shorthandedGoals"],
                stat["shorthandedPoints"],
                nil, #!!!convertStringToTime(stat["stat"]["shortHandedTimeOnIce"]),
                stat["gameWinningGoals"],
                stat["otGoals"],
                nil, #!!!stat["league"]["id"],
                stat["leagueAbbrev"],
                nil, #!!!stat["team"]["id"],
                stat["teamName"]["default"],
            )
        end
    end

    def insertPlayerArchiveStatGoaler(playerId, playerStats)
        playerStats.each do |stat|
            @insertPLayerStatsGoaler.execute(
                playerId,
                stat["season"],
                stat["gamesPlayed"],
                stat["gamesStarted"],
                stat["otLosses"],
                stat["shutouts"],
                stat["wins"],
                stat["losses"],
                convertStringToTime(stat["timeOnIce"]),
                stat["savePctg"],
                nil, #stat["league"]["id"],
                stat["leagueAbbrev"],
                nil, #stat["team"]["id"],
                stat["teamName"]["default"]
            )
        end
    end

    def getPlayerSeasonStatsForPlayerIdAndSeason(playerId, season=nil, league="NHL")
        stats = Array.new

        query = "SELECT * FROM PlayersStatsArchive WHERE playerId = #{playerId}"

        if !season.nil?
            query += " AND season = \"#{season}\""
        end

        query += " AND leagueName = \"#{league}\""
        query += " ORDER BY season"

        results = @dbClient.query(query)

        results.each do |row|
            stat = PlayerSeasonStats.fromRow(row)
            stats.append(stat)
        end

        return stats
    end

    def getPlayerSeasonStatsGoalerForPlayerIdAndSeason(playerId, season=nil, league="NHL")
        stats = Array.new

        query = "SELECT * FROM PlayersStatsArchiveGoaler WHERE playerId = #{playerId}"

        if !season.nil?
            query += " AND season = \"#{season}\""
        end

        query += " AND leagueName = \"#{league}\""
        query += " ORDER BY season"

        results = @dbClient.query(query)

        results.each do |result|
            stat = PlayerSeasonStatsGoaler.new

            stat.season = result["season"]

            stat.games = result["games"]
            stat.gamesStarted = result["gamesStarted"]

            stat.ot = result["ot"]
            stat.shutouts = result["shutouts"]
            stat.wins = result["wins"]
            stat.losses = result["losses"]

            stat.leagueName = result["leagueName"]
            stat.teamName = result["teamName"]

            stat.victoiresProlongation = result["victoiresProlongation"]
            stat.victoiresFusillade = result["victoiresFusillade"]

            stats.append(stat)
        end

        return stats
    end

    def getPoolRosterForSeason(season,all=false)
        roster = Array.new
        query = "SELECT * FROM PoolDraft WHERE season = \"#{season}\""

        if !all
            query += " AND active = 1"
        end

        results = @dbClient.query(query)

        results.each do |row|
            roster.append(row["playerId"])
        end

        return roster
    end

    def getPoolRoster(season, position, statut, all=false)
        roster = Array.new
        query = "SELECT pd.playerId FROM PoolDraft pd, Players p WHERE pd.playerId = p.id AND season = \"#{season}\""

        if position == 'G'
            query += " AND p.positionCode = \"G\""
        else
            query += " AND p.positionCode != \"G\""
        end

        if !all
            if statut.instance_of? Integer
                query += " AND pd.statut = #{statut}"
            else
                query += " AND pd.statut IN (#{statut.join(",")})"
            end
        end

        results = @dbClient.query(query)

        results.each do |row|
            roster.append(row["playerId"])
        end

        return roster
    end

    def addPlayerStatutTaken(playerId, season)
        result = @insertPoolDraft.execute(
            playerId,
            season,
            2,
            nil,
            nil
        )
    end

    def deletePlayerStatutTaken(playerId, season)
        result = @insertPoolDraft.execute(
            playerId,
            season,
            0,
            nil,
            nil
        )
    end

    def selectPlayerForPool(playerId, season, date, points)
        result = @insertPoolDraft.execute(
            playerId,
            season,
            3,
            date,
            points
        )
    end

    def revokePlayerForPool(playerId, season, date, points)
        result = @insertPoolDraft.execute(
            playerId,
            season,
            4,
            date,
            points
        )
    end

    def getAvailablePlayerStatsSalaryForSeason(season)
        results = @dbClient.query("SELECT * from Players p INNER JOIN PlayersStatsArchive psa ON p.id = psa.playerId INNER JOIN PlayersSalaries sal ON p.id = sal.playerId WHERE p.positionCode != 'G' AND psa.season = \"#{season}\" AND sal.season = \"#{season}\" AND psa.leagueName = \"NHL\" AND p.id NOT IN (SELECT playerId FROM PoolDraft WHERE statut IN (1,2)) ORDER BY lastName, firstName")
        players = Array.new

        results.each do |row|
            player = Player.fromRow(row)
            stat = PlayerSeasonStats.fromRow(row)
            salary = PlayerSalarySeason.fromRow(row)

            poolPlayer = PoolPlayerValues.new(player, stat, salary)
            players.append(poolPlayer)
        end

        # Returning a large number will make them at the end of the list
        return players.sort_by { |player| player.value == 0 ? 999999999999 : player.value }
    end

    def insertPlayerSalary(playerId, season, salary)
        result = @insertPlayerSalary.execute(
            playerId,
            season,
            salary
        )
    end

    def getPlayerSeasonSalary(playerId, season)
        result = @dbClient.query("SELECT * FROM PlayersSalaries WHERE playerId = #{playerId} AND season = #{season}")
        row = result.each.first

        if row != nil
            return PlayerSalarySeason.fromRow(row)
        end

        return nil
    end

    private

    def prepareStatements
        @insertTeam = @dbClient.prepare("REPLACE INTO Teams (id,name,venue,abbreviation,firstYearOfPlay,divisionId,conferenceId,franchiseid,active) VALUES (?,?,?,?,?,?,?,?,?)")
        @insertTeamPlayers = @dbClient.prepare("REPLACE INTO TeamsPlayers (teamId,playerId) VALUES (?,?)")
        @insertPosition = @dbClient.prepare("REPLACE INTO Positions (code,abbrev,fullName,type) VALUES (?,?,?,?)")
        @insertPlayer = @dbClient.prepare("REPLACE INTO Players (id,firstName,lastName,primaryNumber,birthYear,birthMonth,birthDay,birthCity,birthProvince,birthCountry,height,weight,active,shoot,rookie,teamId,positionCode,headshotUrl,heroImageUrl) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)")
        @insertPlayerStats = @dbClient.prepare("REPLACE INTO PlayersStatsArchive (playerId,season,games,goals,assists,points,shots,hits,timeOnIce,shifts,plusMinus,shotPct,penaltyMinutes,powerPlayGoals,powerPlayPoints,powerPlayTimeOnIce,shortHandedGoals,shortHandedPoints,shortHandedTimeOnIce,gameWinningGoals,overTimeGoals,leagueId,leagueName,teamId,teamName) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)")
        @insertPLayerStatsGoaler = @dbClient.prepare("REPLACE INTO PlayersStatsArchiveGoaler (playerId,season,games,gamesStarted,ot,shutouts,wins,losses,timeOnIce,savePercentage,leagueId,leagueName,teamId,teamName) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?)")
        @insertPlayerSalary = @dbClient.prepare("INSERT INTO PlayersSalaries (playerId,season,avv) VALUES (?,?,?)")
        @insertPoolDraft = @dbClient.prepare("REPLACE INTO PoolDraft (playerId,season,statut,dateChanged,poolPointsChanged) VALUES (?,?,?,?,?)")
    end

    def convertStringToTime(value)
        if value != nil
            parts = value.split(":")
            firstPart = parts[0].to_f

            secondPart = parts[1].to_f
            secondsPercent = secondPart / 60

            return firstPart + secondsPercent
        else
            return value
        end
    end
end

File: src/main/kotlin/ca/sebleclerc/hockeydata/services/DatabaseService.kt
package ca.sebleclerc.hockeydata.services

import ca.sebleclerc.hockeydata.helpers.Constants
import ca.sebleclerc.hockeydata.helpers.PoolHelper
import ca.sebleclerc.hockeydata.models.cache.CacheRosterPlayer
import ca.sebleclerc.hockeydata.models.Player
import ca.sebleclerc.hockeydata.models.PlayerSalarySeason
import ca.sebleclerc.hockeydata.models.PlayerSkaterSeason
import ca.sebleclerc.hockeydata.models.PoolDraftStatut
import ca.sebleclerc.hockeydata.models.Season
import ca.sebleclerc.hockeydata.models.Team
import ca.sebleclerc.hockeydata.models.cache.CacheGoalerSeason
import ca.sebleclerc.hockeydata.models.cache.CachePlayer
import ca.sebleclerc.hockeydata.models.cache.CacheSkaterSeason
import ca.sebleclerc.hockeydata.models.fromResult
import ca.sebleclerc.hockeydata.models.fromRow
import java.sql.Connection
import java.sql.DriverManager
import java.sql.Statement
import java.sql.Types
import java.util.Properties

private const val url = "jdbc:mariadb://127.0.0.1:3306/hockeydata"
private const val username = "sleclerc"
private const val password = "sleclerc"

class DatabaseService {
  private val connection: Connection
  private val statement: Statement

  init {
    Class.forName("org.mariadb.jdbc.Driver")

    val props = Properties()
    props.setProperty("user", username)
    props.setProperty("password", password)

    connection = DriverManager.getConnection(url, props)
    statement = connection.createStatement()
  }

  // Teams

  fun getAllTeams(): List<Team> {
    val teams = mutableListOf<Team>()

    val rs = statement.executeQuery("SELECT * FROM Teams")
    while (rs.next()) { teams.add(Team.fromResult(rs)) }

    return teams
  }

  fun getTeamForId(teamId: Int): Team? {
    val rs = statement.executeQuery("SELECT * FROM Teams where id = $teamId")

    return if (rs.next()) Team.fromResult(rs) else null
  }

  fun clearRosters() {
    statement.execute("TRUNCATE TABLE TeamsPlayers")
  }

  fun insertTeamRoster(team: Team, players: List<CacheRosterPlayer>) {
    val prepareAddRoster = connection.prepareStatement("REPLACE INTO TeamsPlayers (teamId,playerId) VALUES (?,?)")

    players.forEach { player ->
      prepareAddRoster.setInt(1, team.id)
      prepareAddRoster.setInt(2, player.id)
      prepareAddRoster.execute()
    }
  }

  fun getRosterForTeam(teamId: Int): List<Int> {
    val players = mutableListOf<Int>()

    val rs = statement.executeQuery("SELECT * FROM TeamsPlayers WHERE teamId = $teamId")
    while (rs.next()) { players.add(rs.getInt("playerId")) }

    return players
  }

  // Players

  fun getAllPlayers(onlyGoalers: Boolean? = null): List<Player> {
    val players = mutableListOf<Player>()
    var query = "SELECT * FROM Players"

    onlyGoalers?.also {
      query += if (it) " WHERE positionCode = 'G'" else " WHERE positionCode != 'G'"
    }
    
    val rs = statement.executeQuery(query)

    while (rs.next()) {
      players.add(Player.fromRow(rs))
    }

    return players
  }

  fun getPlayerForId(playerId: Int): Player? {
    val rs = statement.executeQuery("SELECT * FROM Players WHERE id = $playerId")

    return if (rs.next()) Player.fromRow(rs) else null
  }

  fun insertPlayers(player: CachePlayer) {
    val birthSections = player.birthDate.split('-')
    val birthYear = birthSections[0].toInt()
    val birthMonth = birthSections[1].toInt()
    val birthDay = birthSections[2].toInt()

    val insertPlayer = connection.prepareStatement("REPLACE INTO Players (id,firstName,lastName,primaryNumber,birthYear,birthMonth,birthDay,birthCity,birthProvince,birthCountry,height,weight,active,shoot,rookie,teamId,positionCode,headshotUrl,heroImageUrl) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)")
    insertPlayer.setInt(1, player.playerId)
    insertPlayer.setString(2, player.firstName.default)
    insertPlayer.setString(3, player.lastName.default)
    insertPlayer.setInt(4, player.sweaterNumber ?: 0)
    insertPlayer.setInt(5, birthYear)
    insertPlayer.setInt(6, birthMonth)
    insertPlayer.setInt(7, birthDay)
    insertPlayer.setString(8, player.birthCity?.default)
    insertPlayer.setString(9, player.birthStateProvince?.default)
    insertPlayer.setString(10, player.birthCountry)
    insertPlayer.setInt(11, player.heightInInches)
    insertPlayer.setInt(12, player.weightInPounds)
    insertPlayer.setBoolean(13, player.isActive)
    insertPlayer.setString(14, player.shootsCatches)
    insertPlayer.setBoolean(15, false)
    insertPlayer.setObject(16, player.currentTeamId, Types.INTEGER)
    insertPlayer.setString(17, player.position)
    insertPlayer.setString(18, player.headshot)
    insertPlayer.setString(19, player.heroImage)
    insertPlayer.execute()
  }

  fun getSeasonsForSkaterId(playerId: Int): List<PlayerSkaterSeason> {
    val seasons = mutableListOf<PlayerSkaterSeason>()
    val rs = statement.executeQuery("SELECT * FROM PlayersStatsArchive WHERE leagueName = 'NHL' AND gameTypeId = 2 AND playerId = $playerId ORDER BY season DESC LIMIT 5")

    while (rs.next()) {
      seasons.add(PlayerSkaterSeason.fromRow(rs))
    }

    return seasons
  }

  fun insertSkaterSeason(player: CachePlayer, stat: CacheSkaterSeason) {
    val poolPoints = PoolHelper.getSkaterPoolPoint(player, stat)

    val insertStats = connection.prepareStatement("REPLACE INTO PlayersStatsArchive (playerId,season,games,goals,assists,points,shots,hits,timeOnIce,shifts,plusMinus,shotPct,penaltyMinutes,powerPlayGoals,powerPlayPoints,powerPlayTimeOnIce,shortHandedGoals,shortHandedPoints,shortHandedTimeOnIce,gameWinningGoals,overTimeGoals,leagueId,leagueName,teamId,teamName,gameTypeId,poolPoints) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)")
    insertStats.setInt(1, player.playerId)
    insertStats.setInt(2, stat.season)
    insertStats.setObject(3, stat.gamesPlayed, Types.INTEGER)
    insertStats.setObject(4, stat.goals, Types.INTEGER)
    insertStats.setObject(5, stat.assists, Types.INTEGER)
    insertStats.setObject(6, stat.points, Types.INTEGER)
    insertStats.setObject(7, stat.shots, Types.INTEGER)
    insertStats.setNull(8, Types.INTEGER) // Hits
    insertStats.setObject(9, stat.averageTotalOnIce, Types.FLOAT)
    insertStats.setNull(10, Types.INTEGER) // Shifts
    insertStats.setObject(11, stat.plusMinus, Types.INTEGER)
    insertStats.setObject(12, stat.shootingPctg, Types.INTEGER)
    insertStats.setObject(13, stat.pim, Types.INTEGER)
    insertStats.setObject(14, stat.powerPlayGoals, Types.INTEGER)
    insertStats.setObject(15, stat.powerPlayPoints, Types.INTEGER)
    insertStats.setNull(16, Types.FLOAT) // PowerPlayTimeOnIce
    insertStats.setObject(17, stat.shorthandedGoals, Types.INTEGER)
    insertStats.setObject(18, stat.shorthandedPoints, Types.INTEGER)
    insertStats.setNull(19, Types.FLOAT) // ShorthandedTimeOnIce
    insertStats.setObject(20, stat.gameWinningGoals, Types.INTEGER)
    insertStats.setObject(21, stat.otGoals, Types.INTEGER)
    insertStats.setNull(22, Types.INTEGER) // leagueID
    insertStats.setString(23, stat.leagueAbbrev)
    insertStats.setNull(24, Types.INTEGER) // teamID
    insertStats.setString(25, stat.teamName.default)
    insertStats.setInt(26, stat.gameTypeId)
    insertStats.setFloat(27, poolPoints)
    insertStats.execute()
  }

  fun insertGoalerSeason(playerId: Int, stat: CacheGoalerSeason) {
    val insertStats = connection.prepareStatement("REPLACE INTO PlayersStatsArchiveGoaler (playerId,season,games,gamesStarted,ot,shutouts,wins,losses,timeOnIce,savePercentage,leagueId,leagueName,teamId,teamName,gameTypeId) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)")
    insertStats.setInt(1, playerId)
    insertStats.setInt(2, stat.season)
    insertStats.setObject(3, stat.gamesPlayed, Types.INTEGER)
    insertStats.setObject(4, stat.gamesStarted, Types.INTEGER)
    insertStats.setObject(5, stat.otLosses, Types.INTEGER)
    insertStats.setObject(6, stat.shutouts, Types.INTEGER)
    insertStats.setObject(7, stat.wins, Types.INTEGER)
    insertStats.setObject(8, stat.losses, Types.INTEGER)
    insertStats.setObject(9, stat.averageTotalOnIce, Types.FLOAT)
    insertStats.setObject(10, stat.savePctg, Types.FLOAT)
    insertStats.setNull(11, Types.INTEGER) // league ID
    insertStats.setString(12, stat.leagueAbbrev)
    insertStats.setNull(13, Types.INTEGER) //team id
    insertStats.setString(14, stat.teamName.default)
    insertStats.setInt(15, stat.gameTypeId)
    insertStats.execute()
  }

  // Salary

  fun getPlayerSeasonSalary(season: Season, playerId: Int): PlayerSalarySeason? {
    val rs = statement.executeQuery("SELECT * FROM PlayersSalaries WHERE playerId = $playerId AND season = ${season.intValue}")

    return if(rs.next()) PlayerSalarySeason.fromRow(rs) else null
  }

  fun insertPlayerSalary(playerId: Int, season: Season, salary: Int) {
    val insertSalary = connection.prepareStatement("INSERT INTO PlayersSalaries (playerId,season,avv) VALUES (?,?,?)")
    insertSalary.setInt(1, playerId)
    insertSalary.setInt(2, season.intValue)
    insertSalary.setInt(3, salary)
    insertSalary.execute()
  }

  // Pool

  fun getAllPoolDraftStatuses(): Map<Int, PoolDraftStatut> {
    val statuses = mutableMapOf<Int, PoolDraftStatut>()

    val rs = statement.executeQuery("SELECT * FROM PoolDraft WHERE season = ${Constants.currentSeason.intValue}")

    while (rs.next()) {
      val playerId = rs.getInt("playerId")
      val statutId = rs.getInt("statut")
      val status = PoolDraftStatut.values().firstOrNull { it.value == statutId } ?: PoolDraftStatut.AVAILABLE
      statuses[playerId] = status
    }

    return statuses
  }
}
